src/content/post/fastrace-a-modern-approach-to-distributed-tracing-in-rust.mdx

```diff
@@ -0,0 +1,374 @@
+
+```toml
+[dependencies]
+fastrace = "0.7"  # No 'enable' feature
+```
+
+When an application uses this library and doesn't enable the 'enable' feature of fastrace:
+- All tracing code is completely optimized away at compile time
+- Zero runtime overhead is added to the library
+- No impact on performance-critical code paths
+
+When the application does enable tracing via the 'enable' feature:
+- Instrumentation in the dedicated library becomes active
+- Spans are collected and reported
+- The application gets full visibility into library behavior
+
+This is a significant advantage over other tracing solutions that either always impose overhead or require libraries to implement complex feature-flag systems.
+
+### 6. Seamless Context Propagation
+
+`fastrace` provides companion crates for popular frameworks that handle context propagation automatically:
+
+```rust
+// For HTTP clients with reqwest
+let response = client.get(&format!("https://user-service/users/{}", user_id))
+    .headers(fastrace_reqwest::traceparent_headers())
+    .send()
+    .await?;
+
+// For gRPC servers with tonic
+Server::builder()
+    .layer(FastraceServerLayer)  // Automatically extracts context from incoming requests
+    .add_service(MyServiceServer::new(MyService::default()))
+    .serve(addr)
+    .await?;
+
+// For gRPC clients
+let channel = ServiceBuilder::new()
+    .layer(FastraceClientLayer)  // Automatically injects context into outgoing requests
+    .service(channel);
+
+// For data access with OpenDAL
+let op = Operator::new(services::Memory::default())?
+    .layer(FastraceLayer)  // Automatically traces all data operations
+    .finish();
+op.write("test", "0".repeat(16 * 1024 * 1024).into_bytes())
+    .await?;
+```
+
+This provides out-of-box distributed tracing without manual context handling.
+
+## The Complete Solution: `fastrace` + `log` + `logforth`
+
+`fastrace` deliberately focuses on doing one thing well: tracing. Through its composable design and the Rust's great ecosystem, a powerful combination emerges:
+
+- **log**: The standard Rust logging facade
+- **logforth**: A flexible logging implementation with industrial-ready features
+- **fastrace**: High-performance tracing with distributed context propagation
+
+The integration allows your logs to be automatically associated with trace spans, providing correlation without requiring new logging APIs:
+
+```rust
+// Standard log entries are automatically associated with the current span
+log::info!("Processing started");
+
+// Later, in your logging infrastructure, you can see which trace and span
+// each log entry belongs to
+```
+
+To illustrate the simplicity of this approach, here's a streamlined example of building a microservice with complete observability:
+
+```rust
+use fastrace::prelude::*;
+use fastrace_poem::FastraceMiddleware;
+use fastrace_jaeger::JaegerReporter;
+use logforth::diagnostic;
```

Comment:
```suggestion
```

Perhaps inline all the FQN. I can see that `logforth::diagnostic` is used full-qualified below.